hathor/p2p/sync_v2/manager.py

```diff
@@ -0,0 +1,1207 @@
+# Copyright 2023 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import base64
+import json
+import math
+import struct
+from collections import OrderedDict
+from enum import Enum
+from typing import TYPE_CHECKING, Any, Callable, Generator, Optional, cast
+
+from structlog import get_logger
+from twisted.internet.defer import Deferred, inlineCallbacks
+from twisted.internet.task import LoopingCall
+
+from hathor.conf import HathorSettings
+from hathor.p2p.messages import ProtocolMessages
+from hathor.p2p.sync_manager import SyncManager
+from hathor.p2p.sync_v2.mempool import SyncMempoolManager
+from hathor.p2p.sync_v2.streamers import DEFAULT_STREAMING_LIMIT, BlockchainStreaming, StreamEnd, TransactionsStreaming
+from hathor.transaction import BaseTransaction, Block, Transaction
+from hathor.transaction.base_transaction import tx_or_block_from_bytes
+from hathor.transaction.exceptions import HathorError
+from hathor.transaction.storage.exceptions import TransactionDoesNotExist
+from hathor.types import VertexId
+from hathor.util import Reactor, collect_n
+
+if TYPE_CHECKING:
+    from hathor.p2p.protocol import HathorProtocol
+
+settings = HathorSettings()
+logger = get_logger()
+
+MAX_GET_TRANSACTIONS_BFS_LEN: int = 8
+
+
+class PeerState(Enum):
+    ERROR = 'error'
+    UNKNOWN = 'unknown'
+    SYNCING_BLOCKS = 'syncing-blocks'
+    SYNCING_TRANSACTIONS = 'syncing-transactions'
+    SYNCING_MEMPOOL = 'syncing-mempool'
+
+
+class NodeBlockSync(SyncManager):
```

Comment:
I feel it should mostly handle the commands. We can move the syncing logic to another class just like we did with the mempool part.

Reply:
I think it would be a better organization to split it up, but I'd rather avoid splitting up the class right now, mainly because it will cause annoying conflicts on a few other branches that use sync-v2 as base.

Reply:
We will do it on a separate PR to be merged after this one.

Reply:
Created issue: https://github.com/HathorNetwork/hathor-core/issues/728